File: softeer/seats.py
# https://softeer.ai/practice/info.do?eventIdx=1&psProblemId=625

import sys
import math
read = sys.stdin.readline

def cal_dist(one,two):
    x1,y1 = one
    x2,y2 = two
    return math.sqrt((x1-x2)**2 + (y1-y2)**2)

def get_seat(graph, ID):
    
    n = len(graph)
    m = len(graph[0])
    if sum([sum(i) for i in graph]) == 0:
        return [0,0]
    
    can_seat = []
    not_seat = []
    drow = [1,-1,0,0]
    dcol = [0,0,1,-1]
    for row in range(n):
        for col in range(m):
            if graph[row][col] != 0:
                not_seat.append([row,col])
            else:
                can_seat.append([row,col])
            
    cand=[]
    for can in can_seat:
        row = can[0]
        col = can[1]
        possible = [0]*4
        for i in range(4):
            nrow = row + drow[i]
            ncol = col + dcol[i]
            if nrow<0 or nrow>=n or ncol<0 or ncol>=m:
                possible[i] = 1
            
            elif 0<=nrow<n and 0<=ncol<m and graph[nrow][ncol] == 0:
                possible[i] = 1
                                    
        if sum(possible) == 4:
            safety = 10e9
            for seated in not_seat:
                safety = min(safety, cal_dist(can,seated))

            cand.append([row,col,safety])
    
    if len(cand) == 0:
        return False
    return sorted(cand, key = lambda x : (-x[2], x[0], x[1]))[0][:2]
    
def solution(n,m,q,state):
    
    graph = [[0]*m for _ in range(n)]
    eat_check = [0] * 10001
    sitting_check = [0] * 10001
    loc = [0] * 10001
    ans = []
    for how, ID in state:
        ID = int(ID)
        if how == 'In':
            if sitting_check[ID] == 1:
                ans.append(f"{ID} already seated.")

            elif sitting_check[ID] == 0 and eat_check[ID] == 1:
                ans.append(f"{ID} already ate lunch.")
            
            elif sitting_check[ID] == 0 and eat_check[ID] == 0:
                try:
                    x,y = get_seat(graph,ID) 
                    loc[ID] = [x,y]
                    sitting_check[ID] = 1
                    eat_check[ID] = 1
                    graph[x][y] = ID
                    ans.append(f"{ID} gets the seat ({x+1}, {y+1}).")

                except TypeError:
                    ans.append("There are no more seats.")
        
        else:
            if eat_check[ID] == 0:
                ans.append(f"{ID} didn't eat lunch.")
            elif eat_check[ID] == 1 and sitting_check[ID] == 0:
                ans.append(f"{ID} already left seat.")
            elif eat_check[ID] == 1 and sitting_check[ID] == 1:
                x,y = loc[ID]
                ans.append(f"{ID} leaves from the seat ({x+1}, {y+1}).")
                sitting_check[ID] = 0
                graph[x][y] = 0

    return ans

if __name__ =="__main__":
    n,m,q = list(map(int, read().split()))
    state = [list(map(str, read().split())) for _ in range(q)]
    for i in solution(n,m,q,state):
        print(i)

File: softeer/seats.cpp
#include<iostream>
#include<vector>
#include<string>
#include<algorithm>
#include<cmath>
#include<map>

using namespace std;

int matrix[21][21] = {0, };
int drow[4] = {1,-1,0,0};
int dcol[4] = {0,0,1,-1};
int check[10001] ={0,};
map<int, pair<int,int>> now_seat;

int get_safety(int row, int col,int n, int m){
    int ans = 1e9;
    for(int x = 1; x<=n; x++){
        for(int y=1;y<=m;y++){
            if(matrix[x][y] == 1){
                int safety = sqrt(pow(row-x,2) + pow(col-y,2));
                if(safety<ans){ans=safety;}
            }
        }
    }
    return ans;
}

pair<int,int> get_seat(int n,int m){

    pair<int,int> ans;
    bool zero = true;
    for(int row = 1; row<=n; row++){
        for(int col = 1; col<=m; col++){
            if(matrix[row][col] == 1){
                zero = false;
                break;
            }
        }
    }

    if(zero){
        ans.first = 1;
        ans.second = 1;
        return ans;
    }

    vector<pair<pair<int,int>, int>> cand;

    int MAX_safety = -1;
    for(int row = 1; row<=n; row++){
        for(int col = 1; col<=m; col++){
            if(matrix[row][col] == 0){
                bool seat_check = true;
                for(int i =0; i<4; i++){
                    int nrow = row + drow[i];
                    int ncol = col + dcol[i];
                    if(nrow<1 || nrow>n || ncol<1 || ncol>m){continue;}
                    if(matrix[nrow][ncol] == 1){
                        seat_check = false;
                        break;
                        }else{continue;}
                    }
                if(seat_check){
                    int safety = get_safety(row,col,n,m);
                    if(safety > MAX_safety){
                        MAX_safety = safety;
                    }
                    cand.push_back({{row,col}, safety});
                }
            }
        }
    }

    if(cand.size() == 0){
        ans.first = 0;
        ans.second = 0;
        return ans;
    }

    vector<pair<int, int>> new_cand;
    for(int i =0; i<cand.size(); i++){
        if(cand[i].second == MAX_safety){
            new_cand.push_back({cand[i].first.first, cand[i].first.second});
        }
    }

    if(new_cand.size()>1){
        sort(new_cand.begin(), new_cand.end());
    }
    return new_cand[0];
}

int main(){

    int n,m,q;
    cin>>n>>m>>q;
    string state;
    int id;
    vector<pair<string, int>> work;
    for(int i = 0; i<q; i++){
        cin >> state >> id;
        work.push_back({state, id});
    }

    for(int i = 0; i<q; i++){
        state = work[i].first;
        id = work[i].second;

